logstash-core/src/main/java/org/logstash/instrument/metrics/ExtendedFlowMetric.java

```diff
@@ -0,0 +1,360 @@
+    }
+
+    private static void injectIntoRetentionWindows(final List<RetentionWindow> retentionWindows, final FlowCapture capture) {
+        retentionWindows.forEach((rw) -> rw.append(capture));
+    }
+
+    /**
+     * Internal tooling to select the younger of two captures
+     */
+    private static FlowCapture selectNewerCapture(final FlowCapture existing, final FlowCapture proposed) {
+        if (existing == null) { return proposed; }
+        if (proposed == null) { return existing; }
+
+        return (existing.nanoTime() > proposed.nanoTime()) ? existing : proposed;
+    }
+
+    /**
+     * If a window's policy allows it to report before its retention has been reached,
+     * use our lifetime baseline as a default.
+     */
+    private Optional<FlowCapture> windowDefaultBaseline(final RetentionWindow window) {
+        if (window.policy.reportBeforeSatisfied()) {
+            return this.lifetimeBaseline.asOptional();
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Internal introspection for tests to measure how many captures we are retaining.
+     * @return the number of captures in all tracked windows
+     */
+    int estimateCapturesRetained() {
+        return this.retentionWindows.orElse(Collections.emptyList())
+                                    .stream()
+                                    .map(RetentionWindow::estimateSize)
+                                    .mapToInt(Integer::intValue)
+                                    .sum();
+    }
+
+    /**
+     * Internal introspection for tests to measure excess retention.
+     * @param retentionWindowFunction given a policy, return a retention window, in nanos
+     * @return the sum of over-retention durations.
+     */
+    Duration estimateExcessRetained(final ToLongFunction<FlowMetricRetentionPolicy> retentionWindowFunction) {
+        final long currentNanoTime = nanoTimeSupplier.getAsLong();
+        final long cumulativeExcessRetained =
+                this.retentionWindows.orElse(Collections.emptyList())
+                                     .stream()
+                                     .map(s -> s.excessRetained(currentNanoTime, retentionWindowFunction))
+                                     .mapToLong(Long::longValue)
+                                     .sum();
+        return Duration.ofNanos(cumulativeExcessRetained);
+    }
+
+    /**
+     * A {@link RetentionWindow} efficiently holds sufficient {@link FlowCapture}s to
+     * meet its {@link FlowMetricRetentionPolicy}, providing access to the youngest capture
+     * that is older than the policy's allowed retention (if any).
+     * The implementation is similar to a singly-linked list whose youngest captures are at
+     * the tail and oldest captures are at the head, with an additional pre-tail stag.
+     * Compaction is always done at read-time and occasionally at write-time.
+     * Both reads and writes are concurrency-safe.
```

Comment:
NIT: be more explicit about non-blocking intent

```suggestion
     * Both reads and writes are non-blocking and concurrency-safe.
```